### ISSUE TITLE
No definition for max_length in attention

### ISSUE THREAD

**Issue by dyngts** [spro/practical-pytorch#25]
Hi @spro,

I don't know if this is expected, but you don't define the attention size based on max_length of words. Can you explain this? Thank you

**Comment by spro** [comment]
More leftover code... in this case the tutorial [was using](https://github.com/spro/practical-pytorch/blob/0cc55f5aaed44e7903edb8842671411301fcf003/seq2seq-translation/seq2seq-translation.ipynb) a different strategy (Ctrl-F "location-based" in [Effective Approaches to Attention-based Neural Machine Translation](https://arxiv.org/abs/1508.04025)). In that case the attention is calculated solely from the hidden state, using a linear layer with a fixed size output `max_length`.

**Comment by dyngts** [comment]
Hi @spro,

If i can conclude, instead of using location-based this tutorial using content-based, right? so there is no need to define the max_length anymore. How about the evaluation process? we still need the max_length, right?

Thank you

**Comment by spro** [comment]
Right & right. The max length is also used to filter training data.

**Comment by dyngts** [comment]
Got it.

By the way, i have additional question regarding to training process. If i skim your code, the selection of sample training data was done by random choice.
```
training_pair = variables_from_pair(random.choice(pairs))
```

This could lead to some training data will not get selected. How we deal with this?

**Comment by spro** [comment]
There are many more iterations than training examples, so it is fairly likely to cover them all. A more reliable way would be to go through the examples in order and shuffle at the end of every epoch.

Batching would be an improvement but it will take some extra work to implement (the models currently build in the assumption that batch_size=1) - opened #27 to track that.

**Comment by dyngts** [comment]
Thanks @spro,

I see, glad to hear that there will be implementation for batch process. Can't wait for that!
